ntbs-service/DataAccess/NotificationRepository.cs

```diff
@@ -308,15 +313,53 @@ public async Task<IEnumerable<NotificationBannerModel>> GetNotificationBannerMod
                         NotificationStatus = n.NotificationStatus,
                         DrugResistance = n.DrugResistanceProfile.DrugResistanceProfileString,
                         TreatmentEvents = n.TreatmentEvents,
-                        PreviousTbServiceCodes = n.PreviousTbServices.Select(service => service.TbServiceCode),
-                        PreviousPhecCodes = n.PreviousTbServices.Select(service => service.PhecCode),
-                        LinkedNotificationPhecCodes = n.Group.Notifications.Select(no => no.HospitalDetails.TBService.PHECCode),
-                        LinkedNotificationTbServiceCodes = n.Group.Notifications.Select(no => no.HospitalDetails.TBServiceCode)
                     })
-                // Do not use query splitting as EFCore doesn't allow query splitting with a collection in the select
-                .AsSingleQuery()
                 .AsNoTracking()
-                .ToListAsync())
+                .ToListAsync());
+
+            var previousTbServices = _context.Notification
+                .Where(n => ids.Contains(n.NotificationId))
+                .Select(n => new
+                {
+                    n.NotificationId,
+                    PreviousTbServiceCodes = n.PreviousTbServices.Select(service => service.TbServiceCode),
+                    PreviousPhecCodes = n.PreviousTbServices.Select(service => service.PhecCode),
+                })
+                // Do not use query splitting as EFCore doesn't allow query splitting with a collection in the select
+                .AsSingleQuery();
+
+            var groupIds = bannerModels
+                .Select(model => model.GroupId)
+                .Distinct()
+                .ToList();
+
+            var linkedNotifications = _context.NotificationGroup
+                .Where(ng => groupIds.Contains(ng.NotificationGroupId))
+                .Select(ng => new
+                {
+                    GroupId = ng.NotificationGroupId,
+                    LinkedNotificationTbServiceCodes =
+                        ng.Notifications.Select(no => no.HospitalDetails.TBServiceCode),
+                    LinkedNotificationPhecCodes =
+                        ng.Notifications.Select(no => no.HospitalDetails.TBService.PHECCode),
+                })
+                // Do not use query splitting as EFCore doesn't allow query splitting with a collection in the select
+                .AsSingleQuery();
+
+            bannerModels.ForEach(model =>
```

Comment:
A fairly minor thing, but I don't really like using a functional `ForEach` like this - precisely because when you're using `ForEach` you're not really writing functional code (i.e. you're invoking code for its side-effects, rather than what it returns).

In this case, I'd be tempted to transform this into a `Select` by adding a method `CreateNotificationBannerModel` which takes the original model, and also the `previousTbServices` and the `linkedNotifications` to build the fully formed model.

Reply:
Yeah I agree. I've rolled them into the `Select` at the end of the function